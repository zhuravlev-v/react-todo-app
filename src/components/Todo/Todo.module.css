.title {
  font-size: 80px;
  font-weight: 300;
  margin: 0 auto 20px;
  line-height: 1em;
  width: max-content;
}

.todoWrapper {
  width: 100%;
  max-width: 650px;
  max-height: 400px;
}

.todo {
  --line-color: rgb(213 213 213);
  background-color: rgb(252, 252, 252);
  border-radius: 5px;
  -webkit-box-shadow: 0px 0px 15px 0px rgba(126, 126, 126, 0.2);
  -moz-box-shadow: 0px 0px 15px 0px rgba(126, 126, 126, 0.2);
  box-shadow: 0px 0px 15px 0px rgba(126, 126, 126, 0.2);
  display: flex;
  flex-direction: column;
  color: var(--grey-color);

  max-width: 650px;
  width: 100%;
  /* height: 400px; */
  max-height: 400px;
  /* min-height: 400px; */
}

.header {
  /* height: 60px; */
  border-bottom: 1px solid var(--line-color);
  padding: 20px 16px;
  box-sizing: border-box;
  display: flex;
  gap: 25px;
}

.todoInput {
  flex-grow: 1;
}

.main {
  transition: all 0.2s ease;
  overflow: hidden;
  flex-grow: 0;
  height: 0;
}

.mainIsOpen {
  height: auto;
  flex-grow: 1;
  overflow-y: auto;
}

.footer {
  /* height: 34px; */
  font-size: 14px;
  padding: 10px 16px;
  border-top: 1px solid var(--line-color);
  display: flex;
  justify-content: space-between;
  /* margin-top: auto; */
}

.leftItems {}

.tabs {}

.tab {
  color: inherit;
  background: transparent;
  border: none;
  cursor: pointer;
  padding: 3px 4px;
}

.tabActive {
  border: 1px solid #d6d6d6;
  border-radius: 3px;
  /* box-shadow: 1px 1px 15px 0px rgba(126, 126, 126, 0.2); */
}

.clearCompleted {
  color: inherit;
  background: transparent;
  border: none;
  cursor: pointer;
}