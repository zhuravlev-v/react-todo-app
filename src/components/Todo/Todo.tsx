import styles from './Todo.module.css'
import { ITaskObject } from '../../types'
import { useMemo, useState } from 'react'
import TodoAccorderonTrigger from '../TodoAccorderonTrigger/TodoAccorderonTrigger'
import TodoInput from '../TodoInput/TodoInput'
import TodoList from '../TodoList/TodoList'

export default function Todo() {
  const [todoList, setTodoList] = useState([
    { text: 'mvp todo', isDone: true },
    { text: 'add typescript', isDone: true },
    { text: 'add tests', isDone: true },
    { text: 'get job', isDone: false },
  ])

  const [tabs, setTabs] = useState([
    { name: 'all', active: true },
    { name: 'active', active: false },
    { name: 'completed', active: false },
  ])

  const tabActive = tabs.find(tab => tab.active)

  const [accordeonOpen, setAccordeonOpen] = useState(true)

  function handleClickAccordeonTrigger(): void {
    setAccordeonOpen(!accordeonOpen)
  }

  function updateTodoList(newTodoList: Array<ITaskObject>) {
    setTodoList(newTodoList)
  }

  function handleInputEnter(value: string): void {
    addTask(value)
  }

  function addTask(value: string): void {
    const newTodoList = [...todoList]
    newTodoList.push({ text: value, isDone: false })
    setTodoList(newTodoList)
  }

  const leftItemsCounter = todoList.reduce((list, item) => {
    !item.isDone ? list += 1 : null
    return list
  }, 0)

  let leftItemsText = leftItemsCounter > 1 ? leftItemsCounter + ' items left' : leftItemsCounter + ' item left'

  function handleClickTab(index: number) {
    const newTabs = [...tabs]
    newTabs.forEach((tab, tabIndex) => tabIndex === index ? tab.active = true : tab.active = false)
    setTabs(newTabs)
  }

  const sortedTodoList = useMemo(() => {
    if (tabActive?.name === 'active') {
      return [...todoList].filter(task => !task.isDone)
    }
    else if (tabActive?.name === 'completed') {
      return [...todoList].filter(task => task.isDone)
    }
    else {
      return [...todoList]
    }
  }, [tabActive?.name, todoList])

  function clearCompleted(): void {
    const newTodoList = [...todoList].filter(task => task.isDone !== true)
    setTodoList(newTodoList)
  }

  return (
    <div className={styles.todoWrapper} data-testid="todoWrapper">
      <h1 className={styles.title}>Todos</h1>
      <div className={styles.todo}>
        <div className={styles.header}>
          <TodoAccorderonTrigger isOpen={accordeonOpen} onClick={handleClickAccordeonTrigger} />
          <TodoInput className={styles.todoInput} onInputEnter={handleInputEnter} />
        </div>
        <div className={`${styles.main} ${accordeonOpen ? `${styles.mainIsOpen}` : ""}`}>
          <TodoList todoList={sortedTodoList} handleItemChange={updateTodoList} />
        </div>
        <div className={styles.footer}>
          <p className={styles.leftItems}>{leftItemsText}</p>
          <div className={styles.tabs}>
            {
              tabs.map((tab, index) => {
                return (
                  <button
                    className={`${styles.tab} ${tab.active ? `${styles.tabActive}` : ""}`}
                    onClick={() => { handleClickTab(index) }}
                    key={tab.name}
                  >
                    {tab.name}
                  </button>
                )
              })
            }
          </div>
          <button className={styles.clearCompleted} onClick={clearCompleted}>Clear completed</button>
        </div>
      </div>
    </div>
  )
}