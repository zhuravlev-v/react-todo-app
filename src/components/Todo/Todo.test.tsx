import { describe, test, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event'
import Todo from './Todo';

describe("Todo test", () => {
  const user = userEvent.setup()
  // beforeEach(() => {
  //   container = render(<Todo />)
  // });

  test("Component rendered", () => {
    const { container } = render(<Todo />)
    const todoWrapper = container.querySelector('[data-testid="todoWrapper"]')
    expect(todoWrapper).toBeDefined()
  })

  test("Component snapshot", () => {
    const { container } = render(<Todo />)
    expect(container).toMatchSnapshot()
  })

  test('clear completed button rendered', () => {
    render(<Todo />)
    const clearButton = screen.getByText(/Clear completed/i)
    expect(clearButton).toBeDefined()
  })

  test('clear completed button clears completed tasks', async () => {
    render(<Todo />)
    const clearButton = screen.getByText(/Clear completed/i)
    await user.click(clearButton)

    const checkboxes = screen.queryAllByRole('checkbox')

    for (const checkbox of checkboxes) {
      // console.log((checkbox as HTMLInputElement).checked)
      expect((checkbox as HTMLInputElement).checked).not.toBe(true)
    }

  })
})