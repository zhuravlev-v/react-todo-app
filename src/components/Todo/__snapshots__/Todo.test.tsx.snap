// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Todo test > Component snapshot 1`] = `
<div>
  <div
    class="_todoWrapper_5c75ad"
    data-testid="todoWrapper"
  >
    <h1
      class="_title_5c75ad"
    >
      Todos
    </h1>
    <div
      class="_todo_5c75ad"
    >
      <div
        class="_header_5c75ad"
      >
        <button
          class="_button_19efde _isOpen_19efde"
        >
          <img
            alt="open todo list"
            class="_arrow_19efde _arrowIsOpen_19efde"
            src="/arrow-down.svg"
          />
        </button>
        <input
          class="_input_749b1f _todoInput_5c75ad"
          id=""
          name=""
          placeholder="what needs to be done?"
          type="text"
        />
      </div>
      <div
        class="_main_5c75ad _mainIsOpen_5c75ad"
      >
        <ul
          class="_list_5aec6f"
        >
          <li
            class="_listItem_5aec6f"
          >
            <label
              class="_label_00e4ad"
            >
              <input
                checked=""
                class="_checkbox_00e4ad"
                type="checkbox"
              />
              <p
                class="_text_00e4ad _isDone_00e4ad"
              >
                mvp todo
              </p>
            </label>
          </li>
          <li
            class="_listItem_5aec6f"
          >
            <label
              class="_label_00e4ad"
            >
              <input
                checked=""
                class="_checkbox_00e4ad"
                type="checkbox"
              />
              <p
                class="_text_00e4ad _isDone_00e4ad"
              >
                add typescript
              </p>
            </label>
          </li>
          <li
            class="_listItem_5aec6f"
          >
            <label
              class="_label_00e4ad"
            >
              <input
                checked=""
                class="_checkbox_00e4ad"
                type="checkbox"
              />
              <p
                class="_text_00e4ad _isDone_00e4ad"
              >
                add tests
              </p>
            </label>
          </li>
          <li
            class="_listItem_5aec6f"
          >
            <label
              class="_label_00e4ad"
            >
              <input
                class="_checkbox_00e4ad"
                type="checkbox"
              />
              <p
                class="_text_00e4ad "
              >
                get job
              </p>
            </label>
          </li>
        </ul>
      </div>
      <div
        class="_footer_5c75ad"
      >
        <p
          class="_leftItems_5c75ad"
        >
          1 item left
        </p>
        <div
          class="_tabs_5c75ad"
        >
          <button
            class="_tab_5c75ad _tabActive_5c75ad"
          >
            all
          </button>
          <button
            class="_tab_5c75ad "
          >
            active
          </button>
          <button
            class="_tab_5c75ad "
          >
            completed
          </button>
        </div>
        <button
          class="_clearCompleted_5c75ad"
        >
          Clear completed
        </button>
      </div>
    </div>
  </div>
</div>
`;
